api/controllers/UserController.js

```diff
@@ -2559,4 +2559,123 @@ module.exports = {
+   *   '400':
+   *     description: Bad request. See response body for details.
+   *   '401':
+   *     description: Requesting user lacks permission to view requested user.
+   *   '404':
+   *     description: Requested user not found.
+   */
+  async revokeOauthClient(request, internalArgs) {
+    let userId, clientId;
+
+    if (internalArgs && internalArgs.userId && internalArgs.clientId) {
+      userId = internalArgs.userId;
+      clientId = internalArgs.clientId;
+    } else {
+      userId = request.params.id;
+      clientId = request.params.client;
+    }
+
+    // Validate presence of userId param.
+    if (!userId) {
+      logger.warn(
+        '[UserController->revokeOauthClient] No userId provided',
+        {
+          request,
+          fail: true,
+        },
+      );
+      throw Boom.badRequest('No userId provided.');
+    }
+
+    // Validate presence of clientId param.
+    if (!clientId) {
+      logger.warn(
+        '[UserController->revokeOauthClient] No clientId provided',
+        {
+          request,
+          fail: true,
+          user: {
+            id: userId,
+          },
+        },
+      );
+      throw Boom.badRequest('No clientId provided.');
+    }
+
+    // Look up user from DB.
+    const user = await User.findOne({ _id: userId });
+
+    // Validate that user was found in DB.
+    if (!user) {
+      logger.warn(
+        `[UserController->revokeOauthClient] No user found with id ${userId}`,
+        {
+          request,
+          fail: true,
+        },
+      );
+      throw Boom.notFound();
+    }
+
+    // Make sure this OAuth Client exists on the user profile.
+    if (!user.authorizedClients.some(client => client._id.toString() === clientId)) {
+      logger.warn(
+        '[UserController->revokeOauthClient] Requested clientId not found on user profile.',
+        {
+          request,
+          fail: true,
+          user: {
+            id: userId,
+            email: user.email,
+          },
+          oauth: {
+            id: clientId,
+          },
+        },
+      );
+      throw Boom.badRequest('Client ID not found on user profile.');
+    }
+
+    // Validation passed, user exists, client exists on user, so let's remove it.
+    try {
+      const remainingClients = user.authorizedClients.filter(client => client._id.toString() !== clientId);
+      user.authorizedClients = remainingClients;
+      await user.save();
+
+      logger.info(
+        `[UserController->revokeOauthClient] Successfully revoked OAuth Client from user.`,
+        {
+          security: true,
+          user: {
+            id: userId,
+            email: user.email,
+          },
+          oauth: {
+            id: clientId,
+          },
+        },
+      );
+
+      return user;
+    } catch (err) {
+      logger.error(
+        `[UserController->revokeOauthClient] ${err.message}`,
+        {
+          fail: true,
+          stack_trace: err.stack,
+        },
+      );
+    }
```

Comment:
No `Boom` here. Is that fine? Wouldn't that result in no feedback to the user on line 880 of `api/controllers/ViewController.js`?

Reply:
I suppose I never tested that particular code path because all the previous validations prevented me from getting there. I'll add a `throw Boom.internal()` in case there are issues with DB connection or something.

Reply:
I manually wrote `throw Error();` after `await user.save();` and this is what it showed me.

<img width="856" alt="Screen Shot 2021-01-19 at 10 52 09" src="https://user-images.githubusercontent.com/254753/105018151-eae3b680-5a44-11eb-89f2-72552747606b.png">

When we craft other more specific errors, we provide some instructions, but without "real-world" errors I'm not sure what to write, and that will at least be something for now.